meson.build

```diff
@@ -0,0 +1,503 @@
+  'libregexp.c',
+  'libunicode.c',
+  'quickjs.c',
+)
+qjs_hdrs = files(
+  'quickjs.h',
+)
+
+qjs_libc = get_option('libc')
+qjs_libc_srcs = files('quickjs-libc.c')
+qjs_libc_hdrs = files('quickjs-libc.h')
+
+if qjs_libc
+  qjs_srcs += qjs_libc_srcs
+  qjs_hdrs += qjs_libc_hdrs
+
+  qjs_libc_srcs = []
+endif
+
+qjs_c_args = ['-D_GNU_SOURCE']
+
+if host_system == 'windows'
+  qjs_c_args += ['-DWIN32_LEAN_AND_MEAN', '-D_WIN32_WINNT=0x0602']
+endif
+
+qjs_lib = library(
+  'qjs',
+  qjs_srcs,
+
+  # export public headers
+  generator(
+    find_program('cp', 'xcopy'),
+    output: ['@PLAINNAME@'],
+    arguments: ['@INPUT@', '@OUTPUT@'],
+  ).process(qjs_hdrs),
+
+  dependencies: qjs_sys_deps,
+  c_args: qjs_c_args,
+  gnu_symbol_visibility: 'hidden',
+  implicit_include_directories: false,
+  include_directories: include_directories('.'),
+
+  install: true,
+  version: meson.project_version(),
+)
+
+qjs_dep = declare_dependency(
+  link_with: qjs_lib,
+  dependencies: qjs_sys_deps,
+  include_directories: qjs_lib.private_dir_include(),
+)
+
+if host_system == 'emscripten'
+  qjs_wasm_export_name = 'getQuickJs'
+  executable(
+    'qjs_wasm',
+    qjs_srcs,
+    link_args: cc.get_supported_link_arguments(
+      # in emscripten 3.x, this will be set to 16k which is too small for quickjs.
+      '-sSTACK_SIZE=@0@'.format(2 * 1024 * 1024), # let it be 2m = 2 * 1024 * 1024, otherwise, stack overflow may be occured at bootstrap
+      '-sNO_INVOKE_RUN',
+      '-sNO_EXIT_RUNTIME',
+      '-sMODULARIZE', # do not mess the global
+      '-sEXPORT_ES6', # export js file to morden es module
+      '-sEXPORT_NAME=@0@'.format(qjs_wasm_export_name), # give a name
+      '-sTEXTDECODER=1', # it will be 2 if we use -Oz, and that will cause js -> c string convertion fail
+      '-sNO_DEFAULT_TO_CXX', # this project is pure c project, no need for c plus plus handle
+      '-sEXPORTED_RUNTIME_METHODS=ccall,cwrap',
+    ),
+    dependencies: m_dep,
+    c_args: qjs_c_args,
+    implicit_include_directories: false,
+    include_directories: include_directories('.'),
+  )
+endif
+
+install_headers(qjs_hdrs, subdir: 'quickjs-ng')
+
+if not meson.is_subproject()
+  docdir = get_option('docdir')
+  datadir = get_option('datadir')
+
+  if docdir == ''
+    docdir = datadir / 'doc' / meson.project_name()
+  endif
+  install_data(
+    'LICENSE',
+    install_dir: docdir,
+
+    install_tag: 'doc',
+  )
+  install_subdir(
+    'examples',
+    install_dir: docdir,
+
+    strip_directory: false,
+    install_tag: 'doc',
+  )
+endif
+
+meson.override_dependency('quickjs-ng', qjs_dep)
+
+import('pkgconfig').generate(
+  qjs_lib,
+  subdirs: 'quickjs-ng',
+  name: 'quickjs-ng',
+  description: 'QuickJS, the Next Generation: a mighty JavaScript engine',
+  url: 'https://github.com/quickjs-ng/quickjs',
+  version: meson.project_version(),
+)
+
+# QuickJS bytecode compiler
+qjsc_srcs = files(
+  'qjsc.c',
+)
+qjsc_exe = executable(
+  'qjsc',
+  qjsc_srcs,
+  qjs_libc_srcs,
```

Comment:
This will cause all of these sources to be re-compiled for the executable, but then it seems like you link them back in via the `qjs_dep`.

Reply:
`qjs_libc_srcs == []` if `-Dlibc=true`, since it's already compiled in the main `qjs_lib`.
this variable is here to explicitly add `quickjs-libc` to required targets. is it better to use `qjs_libc_lib = static_library(...)` and then add `link_whole: qjs_libc_lib` to `qjs_lib`?

Reply:
applied the `qjs_libc_lib` thing.